src/nvim/getchar.c

```diff
@@ -2050,6 +2050,16 @@ static int vgetorpeek(bool advance)
               save_m_str = vim_strsave(mp->m_str);
               s = eval_map_expr(save_m_str, NUL);
               vgetc_busy = save_vgetc_busy;
+
+              if (State & CMDLINE) {
+                redrawcmdline();
+              } else {
+                if (must_redraw) {
+                  update_screen(0);
+                }
+
+                setcursor();
```

Comment:
any reason not to use the same approach as in https://github.com/vim/vim/commit/4ebe0e62d097d68c5312f9c32714fb41a4c947a3 ? I.e. 

```
int was_screen_col = screen_cur_col;
int was_screen_row = screen_cur_row;

...

ui_cursor_goto(was_screen_row, was_screen_col);
ui_flush();

Reply:
The `update_screen` is needed to clear messages. Super contrived example:

```
function! F1()
    echoerr "test"
    echoerr "test"
    echoerr "test"
    call input("> ")
    return ""
endfunction

map <expr> a F1()
```

Without a screen_update, the cursor is restored, but the screen isn't. We could maybe test for `msg_did_scroll` rather than `must_redraw`, but there could some other obscure corner case that requires a screen update. The ui is flushed the next time we block, or when a something has changed, figured it was fine to wait for that flush.

The cursor strategy does seem better. One thing though, I'm not sure how `ui_cursor_goto` and multigrid interact, I did notice that `ui_cursor_goto` always uses the default grid, would that produce the correct result when multigrid is active? Maybe it takes something like:

```
ui_grid_cursor_goto(curwin->w_grid.handle, was_screen_row, was_screen_col);
```

Reply:
`ui_grid_cursor_goto` is low-level, `ui_cursor_goto` will (must) do the right thing.

Reply:
Great, thanks. I'll make the changes.

I forgot to mention, the `redrawcmdline` call is their to address the bugs I added to the original issue.